Terminal.Gui/ConsoleDrivers/CursesDriver/CursesDriver.cs

```diff
@@ -627,7 +627,7 @@ public override void PrepareToRun (MainLoop mainLoop, Action<KeyEvent> keyHandle
 			return true;
 		});
 
-		_mainLoop.WinChanged += ProcessInput;
```

Comment:
I may be missing something but I always thought `Action = foo" indicated a bug as it meant it would be overwritting someone else's subscription. Which is why I changed it. Does `Action` behave differently in this regard than `event`?

Reply:
You are right and normally we use the incremental subscription for that, but in this case was to avoid add more subscriptions. In pratique the `PrepareToRun` should only run once. The question is if that must be controlled by the `Application` class or by the `ConsoleDriver` class.

Reply:
Just so I'm clear:

- The `action +=` was NOT the bug. The bug is that `PrepareToRun` is being called more than once.
- The correct fix was NOT to do `action =`. 
- The correct fix is to figure out why `PrepareToRun` is being called more than once?
-

Reply:
Found it. Here's the actual bug:

<img width="405" alt="image" src="https://github.com/gui-cs/Terminal.Gui/assets/585482/60a71bd1-ec3b-4ca8-8333-76cc2197247d">

Reply:
Well I think you know why `PrepareToRun` is called every every time a toplevel is started running. It's called by the `Application.Begin` method. I only did in the way it was before. Others solutions are possible. Let `ConsoleDriver` or each derived class handle it.

>Found it. Here's the actual bug:
>
I do not see any bug here. It's only correctly unsubscribing the `ProcessInput` action from the current mainloop instance.

Reply:
> Well I think you know why `PrepareToRun` is called every every time a toplevel is started running. It's called by the `Application.Begin` method. I only did in the way it was before. Others solutions are possible. Let `ConsoleDriver` or each derived class handle it.
> 
> > Found it. Here's the actual bug:
> 
> I do not see any bug here. It's only correctly unsubscribing the `ProcessInput` action from the current mainloop instance.

You're right.

I confused myself. The code to unsubscribe `ChangeWin` is commented out. THAT is a bug.

Sigh. I thought I had figured this out.

Reply:
Ah ok that is another thing :-)

Reply:
> Well I think you know why `PrepareToRun` is called every every time a toplevel is started running. It's called by the `Application.Begin` method. I only did in the way it was before. Others solutions are possible. Let `ConsoleDriver` or each derived class handle it.

`PrepareToRun` is called from `Application.Begin`. It should only be called once per `Application.Begin/End` pair.

If `Application.Begin` does `_mainLoop.ProcessInput += ProcessInput;` and `Application.End` does `_mainLoop.ProcessInput -= ProcessInput;` then `_mainLoop.ProcessInput` should be empty after each `Application.Begin/End` pair. 

I am trying to understand why changing `_mainLoop.ProcessInput += ProcessInput` to `_mainLoop.ProcessInput = ProcessInput` is the correct fix to this issue.

Reply:
> I am trying to understand why changing `_mainLoop.ProcessInput += ProcessInput` to `_mainLoop.ProcessInput = ProcessInput` is the correct fix to this issue.

Simple, because it's changing the current value and not increment a new value. Thus it have always only one subscription.

Reply:
I now realize my mistake. I figured it out as I was falling asleep last night.

I was not considering that `Begin/End` can be called within another `Begin/End` pair. You wrote words above to this effect, but I didn't parse them correctly. Sorry about that.

Changing `+=` to `=` is a workaround, not the fix.

As you noted (I think), the correct fix is to move the setting of `mainloop.ProcessInput from within `Begin` to within `Init` (and the clearing of it from `End` to `Shutdown`. 

RIght?

Reply:
> I now realize my mistake. I figured it out as I was falling asleep last night.
> 
That also happenings to me a lot :-)

> I was not considering that `Begin/End` can be called within another `Begin/End` pair. You wrote words above to this effect, but I didn't parse them correctly. Sorry about that.
> 
That's it. Each toplevel is running with own `Begin/End` pair.

> Changing `+=` to `=` is a workaround, not the fix.
> 
Yes you're right. Only to clarify, when you declare only `Action foo` you can assign either with `+=` or `=`, but when you declare as `event Action foo` you can only assign with `+=`. So if it's already subscribed and you only want one subscription then you need to unsubscribe first with 'foo -= myMethod'.

> I now think the right fix is to change both Windows and Curses' MainLoop implementations to do this:
> 
> <img alt="image" width="478" src="https://user-images.githubusercontent.com/585482/276225375-8230d5d9-42e2-4d55-bbe6-696145387215.png">
> IOW, get rid of the Action all together and just have them call `ProcessInput` on the driver directly.
> 
> I tested it and it appears to work fine. Am I missing something (I probably am).

That a good idea, unless a user want a personalized subscription to handle multi subscriptions based on different mainloops, but isn't the case. Good job.